access_vectors: Add new capabilities to cap2

zpytela opened the pull request:
Updated location of capability definitions to point to current location within kernel source code.

CAP_BPF and CAP_PERFMON mainlined in: cb8e59cc87201af93dfbb6c3dccc8fcad72a09c2, original commit: a17b53c4a4b55ec322c132b6670743612229ee9c
CAP_CHECKPOINT_RESTORE mainlined in: 74858abbb1032222f922487fd1a24513bbed80f9, original commit: 124ea650d3072b005457faed69909221c2905a1f

The missing capabilities were noticed on archlinux with kernel 5.8.14-arch1-1.

Signed-off-by: Dannick Pomerleau <dannickp@hotmail.com>

Resolves: rhbz#1915264

shammancer committed:
access_vectors: Add new capabilities to cap2

Updated location of capability definitions to point to current location within kernel source code.

CAP_BPF and CAP_PERFMON mainlined in: cb8e59cc87201af93dfbb6c3dccc8fcad72a09c2, original commit: a17b53c4a4b55ec322c132b6670743612229ee9c
CAP_CHECKPOINT_RESTORE mainlined in: 74858abbb1032222f922487fd1a24513bbed80f9, original commit: 124ea650d3072b005457faed69909221c2905a1f

The missing capabilities were noticed on archlinux with kernel 5.8.14-arch1-1.

Signed-off-by: Dannick Pomerleau <dannickp@hotmail.com>

Resolves: rhbz#1915264

zpytela committed:
Revert "Allow ptp4l_t sys_admin capability to run bpf programs"

This reverts commit 15b3cef9562733612187b871a1e9bb4dc515dabd.

zpytela committed:
Allow ptp4l_t bpf capability to run bpf programs

In ptp4l, setsockopt() with the SO_ATTACH_FILTER option raises
sk_attach_filter() to run a bpf program.

Running a bpf program requires the BPF capability if it is defined
in selinux-policy, or SYS_ADMIN capability otherwise. This patch
makes the change to use bpf to align with the policy changes.

WOnder93 commented:
We'll also need to add the `perfmon` capability to all domains that use `perf_event`. There is currently a fallback to CAP_SYS_ADMIN, but we shouldn't rely on it long-term. It may be possible to remove `sys_admin` afterwards from some of these domains, but that would need careful testing (or at least verifying that the capability was added specifically to allow something that is now gated by CAP_PERFMON).

WOnder93 commented:
And related to #538, according to the selinux-testsuite policy, `lockdown { confidentiality }` is also needed in case of some `perf_event` related operations (I think only with `cpu` and `tracepoint`, but it will need double-checking).

zpytela committed:
Add perfmon capability for all domains using perf_event

All domains which are allowed to use the perf_event class also need
to have the perfmon capability. Although there is a temporary fallback
check for the sys_admin capability in kernel, it can be dropped in the
near future. With the new perfmon capability in selinux-policy the
primary check is for perfmon.

zpytela commented:
The perfmon cap has been added.

Checking the changed files, no one seems to be a good candidate for removing sys_admin: it is only present in pcp, added with
5bfcae9780e5176d74d61231d1b2af1650ead09e, so unlikely related to perf actions.

zpytela commented:
Lockdown confidentiality added, too.

zpytela commented:
Wait, it should rather go to #538.

WOnder93 commented on policy/modules/kernel/domain.te:
```diff
@@ -245,6 +245,7 @@ optional_policy(`
 
 # Use bpf tools
 allow unconfined_domain_type self:bpf { map_create map_read map_write prog_load prog_run };
+allow unconfined_domain_type self:capability2 perfmon;
```
This should be unnecessary thanks to:
https://github.com/fedora-selinux/selinux-policy/blob/46d834991b9d07400b9feb8aee2d85e40cfe95d7/policy/modules/system/unconfined.if#L23

zpytela replied:
Fixed.

WOnder93 commented:
I think we should consider creating some semantic interfaces in `kernel.if` or `misc_patterns.spt` for BPF and perf_event. For BPF it would be something like:
```
interface(`kernel_use_bpf',`
    allow $1 self:capability2 bpf;
    allow $1 self:bpf { map_create map_read map_write prog_load prog_run };
')
```

Note that for legacy socket filters only the capability is needed - we could have another interface/macro for that. Such a single-rule interface might seem a bit awkward, but it would clearly document the facility that was the reason to add the permission. I believe we should generally move more and more in the direction of semantic macros, otherwise we'll never get the policy to a maintainable state.

And for perf_event:
```
interface(`kernel_use_perf_event',`
    allow $1 self:capability2 perfmon;
    allow $1 self:perf_event open;
')
```
(The additional `perf_event` permissions could be added explicitly as needed or via further interfaces, e.g. `kernel_use_perf_event_all()`.)

zpytela commented:
I agree semantic macros and attributes make the policy better maintainable. Thank you for your feedback, I will get back to the suggestion later.
I actually consider kernel_write_perf_event and kernel_manage_perf_event, based on the current state in the policy.
The bpf class possibly can also make use of some bpf_all_perms.
